### ISSUE TITLE
Unit testing

### ISSUE THREAD

**Issue by Indigo744** [neveldo/jQuery-Mapael#134]
Hello,
I wonder if we should add some UT to Mapael? That would make the development easier?
A lot of people seems to like QUnit (https://qunitjs.com). Never tried it, but seems good.

Here is some reading:
http://andymatthews.net/read/2012/05/29/Unit-Testing-jQuery-Plugins-with-Grunt-and-QUnit
http://javascriptplayground.com/blog/2012/04/a-jquery-plugin-with-grunt-qunit

**Comment by neveldo** [comment]
Hello,

I think that Unit tests are always a good practice whatever is the project. It should indeed help us to prevent code regression with the further changes.

However I'm not familar with unit testing with javascript, I'm wondering where to begin with Mapael, and if the functions could be easily testable. In any case, feel free to make proposals for unit testing Mapael if you have enough time !

**Comment by Indigo744** [comment]
I will try starting something (like instanciating QUnit and building some basic tests). We will build incrementally from that :-)

**Comment by neveldo** [comment]
It's ok for me, thank you for your help !

**Comment by Indigo744** [comment]
I found some example of jQuery plugin using UT:
https://github.com/tabalinas/jsgrid
https://github.com/t4t5/sweetalert

I think we should have only one HTML file, with multiple JS files organized as module (http://api.qunitjs.com/QUnit.module). 
I suggest to sort these different modules regarding the features:
- **Basic**: test default behavior
- **Extended**: test with plots/area
- **Legends**: test everything regarding the legend
- **Zoom event**: test everything regarding the zoom event
- **Update event**: test everything regarding the update event

**Comment by neveldo** [comment]
Thank you, it looks ok to me, I think these modules should cover all the major features of Mapael !

**Comment by Indigo744** [comment]
At the end, we will also be able to automate these tests using Travis or equivalent!
This kind of tools run each time a PR or commit is made, and tells if the build passed or not (by running the Unit Test suite). You should be able to do this from your GitHub control center.

**Comment by neveldo** [comment]
Yes this is a very good point as it will contribute to improve the overall quality of the plugin. I will check the integration of travis through GitHub asap.

**Comment by neveldo** [comment]
I have started to configure automated test (jshint and qunit) through Travis CI : 

- Added Qunit to Grunt config : https://github.com/neveldo/jQuery-Mapael/commit/753d2b55febe0e2c510b7ef602b607d8a5cae7f2
- Added grunt-contrib-qunit to dev dependencies : https://github.com/neveldo/jQuery-Mapael/commit/59e56923b9f609c133b34996233ecfc3d12e08c8
- Added .travis.yml file : https://github.com/neveldo/jQuery-Mapael/commit/a7b59768a464707fec26db2e56540e65cf611adb
- Added travis build information into Readme.md file : https://github.com/neveldo/jQuery-Mapael/commit/bd6c7cfbf966ba1108556639c712059b9d1049f0

Build info on Travis : https://travis-ci.org/neveldo/jQuery-Mapael .

It seems there are two remaining fails. However, I get a weird behavior on my machine : each time I run Qunit through `grunt test`, I get a random number of failed tests ...

**Comment by Indigo744** [comment]
Woah! So nice!
I think there are some trouble with the `setTimeout` functions. I got some errors sometimes, and it would pass after some refresh.
I think this is not the proper way to test this behavior. I will look into it and try another approach.

**Comment by Indigo744** [comment]
We should also integrate a Code Coverage tool in order to monitor our progress on Unit Testing. What do you think?
It exists https://coveralls.io or https://codecov.io which are free for Open Source project.

**Comment by neveldo** [comment]
Hello,

I agree with you, it's a great idea ! I think I will set up coveralls.io, it seems that it is a lot more documented than codecov.io.

**Comment by Indigo744** [comment]
Great :) let me know when it is ready!
